org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java

```diff
@@ -214,11 +218,17 @@ public void runInToolingApi(CancellationTokenSource tokenSource, IProgressMonito
                 SubMonitor progress = SubMonitor.convert(monitor, 5);
                 progress.setTaskName((String.format("Synchronizing Gradle build at %s with workspace", this.gradleBuild.getBuildConfig().getRootProjectDirectory())));
                 new ImportRootProjectOperation(this.gradleBuild.getBuildConfig(), this.newProjectHandler).run(progress.newChild(1));
-                Set<EclipseProject> allProjects = ModelProviderUtil.fetchAllEclipseProjects(this.gradleBuild, tokenSource, FetchStrategy.FORCE_RELOAD, progress.newChild(1));
+                Collection<EclipseProject> rootProjects = ModelProviderUtil.fetchAllEclipseProjectsWithSyncTask(this.gradleBuild, tokenSource, FetchStrategy.FORCE_RELOAD, progress.newChild(1));
+                Set<EclipseProject> allProjects = collectAll(rootProjects);
                 new ValidateProjectLocationOperation(allProjects).run(progress.newChild(1));
                 new RunOnImportTasksOperation(allProjects, this.gradleBuild.getBuildConfig()).run(progress.newChild(1), tokenSource);
                 this.failures = new SynchronizeGradleBuildOperation(allProjects, this.gradleBuild, this.newProjectHandler,
                         ProjectConfigurators.create(this.gradleBuild, CorePlugin.extensionManager().loadConfigurators())).run(progress.newChild(1));
+
+                // The root projects are no longer loaded via GradleBuild.getModelProvider(). To avoid unnecessary
+                // model loads (e.g. when somebody opens the test sources from the Execution view) we add the result
+                // to the cache here.
+                ((DefaultModelProvider)this.gradleBuild.getModelProvider()).getCache().put(EclipseProject.class, rootProjects);
```

Comment:
❌ Why do we have to do this here, when you've added the caching to the phased actions too?

Reply:
Yeah, now we have two types of model caches. One that kicks in only during synchronization and happens automatically and another one that is long-lived (i.e. `ModelProvider`). Because we moved away from `ModelProvider` usage for loading the Eclipse model, other code pieces like the `Open Test Sources` functionality would have to load the same Eclipse model as it was retrieved during the sync.

Reply:
Any way we could unify them and have everything be long-lived? I really don't like exposing the internals here like this.

Reply:
I don't think that's feasible, at least not with a lot of effort. Alternatively, we could handle the caching with implementing a `ModelProvider.queryEclipseModelAndExecuteTasks()` method. 

`ModelProvider` and `GradleBuild.withConnection` has duplicate functionality, that's the real issue here.

Reply:
If it would be a lot of effort then I'd be in favor of a new method instead.

Reply:
I've moved the query to the `ModelProvide` interface.